From 09957cfd27c202594dd512e076b992ce2aed30fe Mon Sep 17 00:00:00 2001
Message-ID: <09957cfd27c202594dd512e076b992ce2aed30fe.1764154288.git.gerald.elder-vass@citrix.com>
In-Reply-To: <61995ad32249580db2c1e899b495a41abfe44d3e.1764154288.git.gerald.elder-vass@citrix.com>
References: <61995ad32249580db2c1e899b495a41abfe44d3e.1764154288.git.gerald.elder-vass@citrix.com>
From: Gerald Elder-Vass <gerald.elder-vass@citrix.com>
Date: Wed, 26 Nov 2025 10:50:29 +0000
Subject: [PATCH 09/12] Revert "hw/i386/pc: Remove
 PCMachineClass::resizable_acpi_blob"

This reverts commit 91616f812a1d0bbca9efb3f87ca128759f01f5f8.
---
 hw/i386/acpi-build.c | 10 ++++++++++
 hw/i386/pc.c         |  1 +
 include/hw/i386/pc.h |  3 +++
 3 files changed, 14 insertions(+)

diff --git a/hw/i386/acpi-build.c b/hw/i386/acpi-build.c
index d0be066feea5..d89bd308082b 100644
--- a/hw/i386/acpi-build.c
+++ b/hw/i386/acpi-build.c
@@ -2119,6 +2119,16 @@ void acpi_build(AcpiBuildTables *tables, MachineState *machine)
      * keep the table size stable for all (max_cpus, max_memory_slots)
      * combinations.
      */
+    /* Make sure we have a buffer in case we need to resize the tables. */
+    if ((tables_blob->len > ACPI_BUILD_TABLE_SIZE / 2) &&
+        !pcmc->resizable_acpi_blob) {
+        /* As of QEMU 2.1, this fires with 160 VCPUs and 255 memory slots.  */
+        warn_report("ACPI table size %u exceeds %d bytes,"
+                    " migration may not work",
+                    tables_blob->len, ACPI_BUILD_TABLE_SIZE / 2);
+        error_printf("Try removing CPUs, NUMA nodes, memory slots"
+                     " or PCI bridges.\n");
+    }
     acpi_align_size(tables_blob, ACPI_BUILD_TABLE_SIZE);
 
     acpi_align_size(tables->linker->cmd_blob, ACPI_BUILD_ALIGN_SIZE);
diff --git a/hw/i386/pc.c b/hw/i386/pc.c
index 32f39147938b..7603031b9b3c 100644
--- a/hw/i386/pc.c
+++ b/hw/i386/pc.c
@@ -1813,6 +1813,7 @@ static void pc_machine_class_init(ObjectClass *oc, const void *data)
     pcmc->acpi_data_size = 0x20000 + 0x8000;
     pcmc->pvh_enabled = true;
     pcmc->kvmclock_create_always = true;
+    pcmc->resizable_acpi_blob = true;
     x86mc->apic_xrupt_override = true;
     assert(!mc->get_hotplug_handler);
     mc->get_hotplug_handler = pc_get_hotplug_handler;
diff --git a/include/hw/i386/pc.h b/include/hw/i386/pc.h
index 44444a268f44..6f85f73baaef 100644
--- a/include/hw/i386/pc.h
+++ b/include/hw/i386/pc.h
@@ -133,6 +133,9 @@ struct PCMachineClass {
     /* create kvmclock device even when KVM PV features are not exposed */
     bool kvmclock_create_always;
 
+    /* resizable acpi blob compat */
+    bool resizable_acpi_blob;
+
     /*
      * whether the machine type implements broken 32-bit address space bound
      * check for memory.
-- 
2.47.3

